### PR TITLE
Source encoding always set when opening datasets

### LINKED ISSUE #2550: Include filename or path in open_mfdataset
When reading from multiple files, sometimes there is information encoded in the filename. For example in these grib files the time: `./ST4.2018092500.01h`, `./ST4.2018092501.01h`. It seems like a generally useful thing would be to allow the passing of a `kwargs` (such as `path_as_coord` or something) that would define a set of coords with one for the data from each file. 

I think the code change would be small:

```python 
if path_as_coord:
    ds = ds.assign_coords(path=file_name)
```

In use it would be like: 

```python
>>>xr.open_mfdataset(['./ST4.2018092500.01h', './ST4.2018092501.01h'], engine='pynio', concat_dim='path')
<xarray.Dataset>
Dimensions:               (x: 881, y: 1121, time: 2)
Coordinates:
    lat             (x, y) float32 23.116999 ... 45.618984
    lon             (x, y) float32 -119.023 ... -59.954613
  * path            (path) <U20 './ST4.2018092500.01h' './ST4.2018092501.01h'
Dimensions without coordinates: x, y
Data variables:
    var_0  (time, x, y) float32 dask.array<shape=(2, 881, 1121), chunksize=(1, 881, 1121)>
    var_1   (time, x, y) float32 dask.array<shape=(2, 881, 1121), chunksize=(1, 881, 1121)>
```

For context I have implemented something similar in dask: https://github.com/dask/dask/pull/3908

### PR DESCRIPTION
**Pull request by TomNicholas**
Closes #2550 by ensuring that the filename is always stored in the encoding dictionary of a dataset, under `'source'`. Previously this feature would be backend-dependent. 

This was motivated by wanting the `preprocess` function passed to `open_mfdataset` to have access to the filename of the dataset it is operating on. A specific use case is when you are opening a spatial grid of datasets, and want to apply a different operation to datasets at the edge of the grid (e.g. to deal with guard cells). It is therefore also relevant for discussion in [#2159](https://github.com/pydata/xarray/issues/2159#issuecomment-391512018).

 - [x] Closes #2550
 - [x] Tests added
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

### CODE REVIEW

File: xarray/backends/api.py [hunk 1]
```diff
@@ -338,17 +338,22 @@ def maybe_decode_store(store, lock=False):
...
-            return maybe_decode_store(store)
     else:
         if engine is not None and engine != 'scipy':
             raise ValueError('can only read file-like objects with '
                              "default engine or engine='scipy'")
         # assume filename_or_obj is a file-like object
         store = backends.ScipyDataStore(filename_or_obj)
 
-    return maybe_decode_store(store)
+    with close_on_error(store):
```
[the remark this answers is not in the document]
Comment (TomNicholas) [reply xarray/backends/api.py]: I did wonder if that was deliberate... I've restored the old behaviour now.

**Comment by dcherian** [comment]
I think the what's new entry could be a little clearer and it would be good to add this to the docs somewhere. Here's one option: http://xarray.pydata.org/en/stable/io.html?highlight=encoding#reading-encoded-data

LGTM otherwise

File: doc/whats-new.rst [hunk 2]
```diff
@@ -61,6 +61,9 @@ Enhancements
 - :py:meth:`DataArray.resample` and :py:meth:`Dataset.resample` now supports the
   ``loffset`` kwarg just like Pandas.
   By `Deepak Cherian <https://github.com/dcherian>`_
+- Datasets are now guaranteed to have a ``'source'`` encoding, so the source
```
[the remark this answers is not in the document]
Comment (shoyer) [reply doc/whats-new.rst]: You can't do attribute lookups in `encoding`, so `encoding.source` isn't valid.

[the remark this answers is not in the document]
Comment (dcherian) [reply doc/whats-new.rst]: Oh right. So `encoding['source']` then?

[the remark this answers is not in the document]
Comment (shoyer) [reply doc/whats-new.rst]: Yep, `encoding['source']` would work

**Comment by shoyer** [comment]
thanks!

**Pull request merged**